README.md
# NAME

Mojolicious::Plugin::Loco - launch local GUI via default web browser

# VERSION

version 0.001

# SYNOPSIS

```perl
# Mojolicious::Lite

plugin 'Loco';

get '/' => "index";
#...
app->start;

__DATA__

@@ index.html.ep
% layout 'default';
#...

@@ layouts/default.html.ep
<!DOCTYPE html>
<html><head>

%= $c->loco->jsload;

</head><body>
%= content
</body></html>
```

# DESCRIPTION

On server start, [Mojolicious::Plugin::Loco](https://metacpan.org/pod/Mojolicious::Plugin::Loco) opens a dedicated window in your default internet browser.  This assumes an available desktop that [Browser::Open](https://metacpan.org/pod/Browser::Open) knows how to deal with.  The application must be listening on a loopback/localhost port (dies otherwise), and the server shuts down when the browser window (and descendants thereof) is subsequently closed.

This is a way to create low-effort desktop applications using [Mojolicious](https://metacpan.org/pod/Mojolicious) (cross-platform if your code is sufficiently portable).

# OPTIONS

## initial\_wait

How many seconds to wait on server start for browser window to finish loading.  Defaults to 10.

## final\_wait

How many seconds to wait after browser window ceases communicating before terminating (we do not rely on `window.unload`).  Defaults to 3.

Since javascript timer events from backgrounded/hidden tabs/windows are typically throttled, reducing this below 2 will most likely make the application terminate whenever the window is hidden or minimized.

## entry\_path

URI path for the entry point of your application (i.e., what to display when the brower window opens).  Defaults to `/`.

## api\_prefix

Path prefix for URIs used by this plugin.  This is where the various endpoints needed by this module live (and is also where the required javascript file(s) are served from).  Defaults to `/hb/`; it can be pretty much anything as long as it's distinct from what the rest of your application uses.  But it would be best to keep it short.

# HELPERS

## loco.jsload

Loads whatever javascript needs to be in the &lt;head> section of every page to be displayed in the browser window.  You most likely want this in your default layout.

```
%= $c->loco->jsload;
```

Or you can be more elaborate

```perl
%= javascript 'https://code.jquery.com/jquery-3.3.1.min.js';
%= $c->loco->jsload( nojquery => 1, begin
      .on_hb(function(h) {
        // do something on every heartbeat
        $('#heartbeat').html(h);
      })
% end );
```

Options include

### nojquery

Suppress loading of jquery from the Mojolicious distribution, meaning you already loaded it from elsewhere (needs to be at least version 3.0).

### nofinish

Suppress the default `on_finish` handler.

### begin

Final `begin` block, if provided, will be assumed to be javascript code to further configure the heartbeat object (code is preceded by `$().heartbeat()`), typically to add `on_hb` or `on_finish` handlers

# METHODS

[Mojolicious::Plugin::Loco](https://metacpan.org/pod/Mojolicious::Plugin::Loco) inherits all methods from
[Mojolicious::Plugin](https://metacpan.org/pod/Mojolicious::Plugin) and implements the following new ones.

## register

```perl
$plugin->register(Mojolicious->new, entry_path => '/');
```

Register plugin in [Mojolicious](https://metacpan.org/pod/Mojolicious) application.

# SEE ALSO

[Mojolicious](https://metacpan.org/pod/Mojolicious), [Mojolicious::Guides](https://metacpan.org/pod/Mojolicious::Guides), [https://mojolicious.org](https://mojolicious.org).

# AUTHOR

Roger Crew <wrog@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 by Roger Crew.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

0.008     Oct 25 2019
        * fix metacpan issues, remove CONTRIBUTING.pod

0.007     Oct 22 2019
        * remove author dependencies from Makefile.PL

0.006     Aug 10 2018
        * loco->conf renamed to loco->config
        * Application config can override option settings
            per new plugin option: config_key
        * All options are now settable

0.005     Aug 06 2018
        * Fixed build/test prerequisites

0.004     Aug 05 2018
        * Moar tests
        * Changes to plugin options
            'browser' - new option
            'allow_other_browsers' -> 'allow_other_sessions'
            'initial_wait' - new default (15), now settable
        * Explicit session id identifes launched browser
            new helpers 'loco->id' and 'loco->id_fail'
        * Fixed non-mswin32 browser launch
        * Removed restrictions on listening points

0.003     Jul 31 2018

0.002     Jul 31 2018
        * loco->jsload now takes 'jquery' option rather than 'nojquery'
        * New config option: allow_other_browsers
        * New helpers: loco->quit, loco->csrf_fail
        * Added SECURITY CONSIDERATIONS
        * Fixed module packaging issues

0.001     Jul 29 2018
        * Initial version


t/MyTest.pm
package MyTest;

use Mojo::Base -strict;
use Browser::Open;

our @browser_open_cb = ();
Mojo::Util::monkey_patch 'Browser::Open', open_browser => sub {
    $_->(@_) for @browser_open_cb;
};

1;


lib/Mojolicious/Plugin/Loco.pm
package Mojolicious::Plugin::Loco;

# ABSTRACT: launch a web browser; easy local GUI

use Mojo::Base 'Mojolicious::Plugin';

use Browser::Open 'open_browser_cmd';
use File::ShareDir 'dist_file';
use Mojo::ByteStream 'b';
use Mojo::Util qw(hmac_sha1_sum steady_time);

sub register {
    my ($self, $app, $o) = @_;
    my $conf = {
        config_key   => 'loco',
        entry        => '/',
        initial_wait => 15,
        final_wait   => 3,
        api_path     => '/hb/',
        %$o,
    };
    if (my $loco = $conf->{config_key}) {
        unless (my $ac = $app->config($loco)) {
            $app->config($loco, $conf);
        }
        else {
            %$ac = (%$conf, %$ac);
            $conf = $ac;
        }
    }

    my $api =
      Mojo::Path->new($conf->{api_path})->leading_slash(1)->trailing_slash(1);
    my ($init_path, $hb_path, $js_path) =
      map { $api->merge($_)->to_string } qw(init hb heartbeat.js);

    $app->helper(
        'loco.config' => sub {
            my $c = shift;

            # Hash
            return $conf unless @_;

            # Get
            return $conf->{ $_[0] } unless @_ > 1 || ref $_[0];

            # Set
            my $values = ref $_[0] ? $_[0] : {@_};
            @{$conf}{ keys %$values } = values %$values;
            return $c;
        }
    );

    $app->helper(
        'loco.reply_400' => sub {
            my $c       = shift;
            my %options = (
                info   => '',
                status => $c->stash('status') // 400,
                (@_ % 2 ? ('message') : (message => 'Bad Request')), @_
            );
            return $c->render(template => 'done', title => 'Error', %options);
        }
    );

    $app->helper(
        'loco.csrf_fail' => sub {
            my $c = shift;
            return 1 if '400' eq ($c->res->code // '');
            return $c->loco->reply_400(info => 'unexpected origin')
              if $c->validation->csrf_protect->error('csrf_token');
        }
    );

    $app->helper(
        'loco.id_fail' => sub {
            my $c = shift;
            return 1 if '400' eq ($c->res->code // '');
            return $c->loco->reply_400(info => 'wrong session')
              unless $c->loco->id;
        }
    );

    $app->helper(
        'loco.quit' => sub {
            my $c = shift;
            return if $c->loco->csrf_fail;
            $c->render(
                template => "done",
                format   => "html",
                title    => "Finished",
                header   => "Close this window"
            ) unless $c->res->code;
            Mojo::IOLoop->timer(1 => sub { shift->stop });
        }
    );

    $app->hook(
        before_server_start => sub {
            my ($server, $app) = @_;
            return if $conf->{browser_launched}++;
            my ($url) = map {
                my $u = Mojo::URL->new($_);
                $u->host($u->host =~ s![*]!localhost!r);
            } @{ $server->listen };

            my $_test = $conf->{_test_browser_launch};

            # no explicit port means this is coming from UserAgent
            return
              unless ($url->port || $_test);

            $conf->{seed} = my $seed =
              _make_csrf($app, $$ . steady_time . rand . 'x');

            $url->path($init_path)->query(s => $seed);

            my $cmd = $conf->{browser} // open_browser_cmd();
            unless ($cmd) {
                die "Cannot find browser to execute"
                  unless defined $cmd;
                return;
            }
            elsif (ref($cmd) eq 'CODE') {
                $cmd->($url);
            }
            else {
                if ($_test) {
                    $_test->($cmd, $url);
                    return;
                }
                if ($^O eq 'MSWin32') {
                    system start => (
                        $cmd =~ m/^microsoft-edge/
                        ? ("microsoft-edge:$url")
                        : (($cmd eq 'start' ? () : ($cmd)), "$url")
                    ) and die "exec '$cmd' failed";
                }
                else {
                    my $pid;
                    unless ($pid = fork) {
                        unless (fork) {
                            exec $cmd, $url->to_string;
                            die "exec '$cmd' failed";
                        }
                        exit 0;
                    }
                    waitpid($pid, 0);
                }
            }
            _reset_timer($conf->{initial_wait});
        }
    );

    $app->hook(
        before_routes => sub {
            my $c = shift;
            $c->validation->csrf_token('')
              if ($conf->{seed} || !$c->session->{'loco.id'});
        }
    ) unless $conf->{allow_other_sessions};

    $app->helper(
        'loco.id' => sub {
            my $c = shift;
            undef $c->session->{csrf_token}
              if @_;
            return $c->session('loco.id', @_);
        }
    );

    $app->routes->get(
        $init_path => sub {
            my $c    = shift;
            my $seed = $c->param('s') // '' =~ s/[^0-9a-f]//gr;

            if (length($seed) >= 40
                && $seed eq ($conf->{seed} // ''))
            {
                delete $conf->{seed};
                $c->loco->id(1);
            }
            $c->redirect_to($conf->{entry});
        }
    );

    $app->routes->get(
        $hb_path => sub {
            my $c = shift;
            state $hcount = 0;
            if ($c->validation->csrf_protect->error('csrf_token')) {

                # print STDERR "bad csrf: "
                # . $c->validation->input->{csrf_token} . " vs "
                # . $c->validation->csrf_token . "\n";
                return $c->render(
                    json    => { error => 'unexpected origin' },
                    status  => 400,
                    message => 'Bad Request',
                    info    => 'unexpected origin'
                );
            }
            _reset_timer($conf->{final_wait});
            $c->render(json => { h => ++$hcount });

            #    return $c->helpers->reply->not_found()
            #      if ($hcount > 5);
        }
    );

    $app->static->extra->{ $js_path =~ s!^/!!r } =
      dist_file(__PACKAGE__ =~ s/::/-/gr, 'heartbeat.js');

    push @{ $app->renderer->classes }, __PACKAGE__;

    $app->helper(
        'loco.jsload' => sub {
            my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
            my ($c, %option) = @_;
            my $csrf   = $c->csrf_token;
            my $jquery = $option{jquery} // '/mojo/jquery/jquery.js';
            b(
                (
                    join "",
                    map { $c->javascript($_) . "\n" }
                      (length($jquery) ? ($jquery) : ()),
                    $js_path
                )
                . $c->javascript(
                    sub {
                        <<END
\$.fn.heartbeat.defaults.ajax.url = '$hb_path';
\$.fn.heartbeat.defaults.ajax.headers['X-CSRF-Token'] = '$csrf';
END
                          . $c->include(
                            'ready',
                            format   => 'js',
                            nofinish => 0,
                            %option, _cb => $cb
                          );
                    }
                )
            );
        }
    );

}

sub _make_csrf {
    my ($app, $seed) = @_;
    hmac_sha1_sum(pack('h*', $seed), $app->secrets->[0]);
}

sub _reset_timer {
    state $timer;
    Mojo::IOLoop->remove($timer)
      if defined $timer;
    return unless my $wait = shift;
    $timer = Mojo::IOLoop->timer(
        $wait,
        sub {
            print STDERR "stopping...";
            shift->stop;
        }
    );
}

1;

__DATA__

@@ layouts/done.html.ep
<!DOCTYPE html>
<html>
  <head><title><%= title %></title>
%= stylesheet begin
body { background-color: #ddd; font-family: helvetica; }
h1 { font-size: 40px; color: white }
% end
  </head>
  <body><%= content %></body>
</html>

@@ done.html.ep
% layout 'done';
% title $title;
<h1 id="header"><%= $header %></h1>
<h2><span id="status"></span> <span id="message"></span></h2>
<p id="info"></p>

@@ ready.js.ep
$.ready.then(function() {
    $().heartbeat()
%== $_cb ? $_cb->() : ''
% unless ($nofinish) {
    .on_finish(function(unexpected,o) {
% my ($hd,$bdy) = do {
%   my $d = Mojo::DOM->new($c->render_to_string(template => "done", format => "html", title => "Finished", header => "Close this window"));
%   map {"'" . (Mojo::Util::trim($d->at($_)->content)
%                =~ s/'/\\047/gr =~ s/\n/'\n    +'\\n/gr) . "'"} qw(head body)
% };
      $('head').html(<%== $hd %>);
      $('body').html(<%== $bdy %>);
      if (unexpected) {
        $('#header').html('Error');
        $('#status').html(o.code);
        $('#message').html(o.msg);
        $('#info').html(o.status == 'error' ? '' : "("+o.msg+")");
      }
    })
% }
    .start();
});


lib/Mojolicious/Plugin/LocalUI.pm
package Mojolicious::Plugin::LocalUI;
# ABSTRACT: launch a dedicated local web browser UI window

use Mojo::Base 'Mojolicious::Plugin';

use Browser::Open 'open_browser';
use File::ShareDir 'dist_file';
use Mojo::ByteStream 'b';
use Mojo::Util qw(hmac_sha1_sum steady_time);

sub register {
    my ($self, $app, $o) = @_;
    my %conf = (
	entry_path => '/',
	initial_wait => 10,
	final_wait => 3,
	api_prefix => '/hb/',
	%$o
       );
    my $api = Mojo::Path->new($conf{api_prefix})
      ->leading_slash(1)->trailing_slash(1);
    my ($init_path, $hb_path, $js_path)
      = map {$api->merge($_)->to_string} 
      qw(init hb heartbeat.js);

    $app->helper( 'localui.conf' => sub { \%conf });
    $app->hook(
	before_server_start => sub {
	    my ($server, $app) = @_;
	    return if $conf{browser_launched};
	    ++$conf{browser_launched};
	    return if (caller(7))[0] =~ m/Command::get$/;
	    my ($url) = 
	      map { 
		  $_->host($_->host =~ s![*]!localhost!r); 
	      }
	      grep {
		  $_->host =~ m/^(?:[*]|localhost|127[.]([0-9.]+))$/
	      }
	      map { Mojo::URL->new($_) }
	      @{$server->listen};
	    die "Must be listening at a loopback URI" unless $url;

	    $conf{seed} =
	      my $seed = _make_csrf($app, $$ . steady_time . rand . 'x');

	    $url->path($init_path)->query(s => $seed);
	    my $e = open_browser($url->to_string);
	    if ($e // 1) {
		unless ($e) {
		    die "Cannot find browser to execute";
		}
		else {
		    die "Error executing: ". Browser::Open::open_browser_cmd . "\n";
		}
	    }
	    _reset_timer($conf{initial_wait});
	});
    $app->routes->get(
	$init_path => sub {
	    my $c = shift;
	    my $seed = $c->param('s')//'' =~ s/[^0-9a-f]//gr;

	    my $u = Mojo::URL->new($conf{entry_path});
	    if (length($seed) >= 40
		  && $seed eq ($conf{seed}//'')) {
		delete $conf{seed};
		undef $c->session->{csrf_token}; # make sure we get a fresh one
		$conf{csrf} = my $csrf = $c->csrf_token;
		$u->query(csrf_token => $csrf);
	    }
	    $c->redirect_to($u);
	});
    $app->routes->get(
	$hb_path => sub {
	    my $c = shift;
	    state $hcount = 0;
	    if ($c->validation->csrf_protect->error('csrf_token')) {
print STDERR "bad csrf: ".$c->validation->input->{csrf_token}." vs ".$c->validation->csrf_token."\n";
		return $c->render(json => { error => 'unexpected origin' },
				  status => 400, message => 'Bad Request', info => 'unexpected origin');
	    }
	    _reset_timer($conf{final_wait});
	    $c->render(json => { h => ++$hcount });
	    #    return $c->helpers->reply->not_found()
	    #      if ($hcount > 5);
	});
    # $app->hook(
    # 	before_dispatch => sub {
    # 	    my $c = shift;
    # 	    return unless $conf{csrf};
    # 	    $c->reply->bad_request(info => 'unexpected origin')
    # 	      if $c->validation->csrf_protect->error('csrf_token');
    # 	});

    $app->static->extra->{$js_path =~ s!^/!!r} = 
      dist_file(__PACKAGE__ =~ s/::/-/gr, 'heartbeat.js');

    push @{$app->renderer->classes}, __PACKAGE__;
    
    $app->helper(
	'localui.jsload' => sub { 
	    my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	    my ($c, %option) = @_; 
	    my $csrf = $c->param('csrf_token')//'missing';
	    b(( join "", map { $c->javascript($_)."\n" }
		($option{nojquery} ? () : ('/mojo/jquery/jquery.js')),
		$js_path 
	       ) .
	       $c->javascript( sub { <<END 
\$.fn.heartbeat.defaults.ajax.url = '$hb_path';
\$.fn.heartbeat.defaults.ajax.headers['X-CSRF-Token'] = '$csrf';
END
 . $c->include('ready', format => 'js', nofinish => 0, %option, _cb => $cb);
}))
	});

    # $app->helper(
    # 	'reply.bad_request', sub {
    # 	    my $c = shift;
    # 	    my %options = (info => '', status => $c->stash('status') // 400,
    # 			   (@_%2 ? ('message') : (message => 'Bad Request')), @_);
    # 	    return $c->render(template => 'done', title => 'Error', %options);
    # });
}

sub _make_csrf {
    my ($app, $seed) = @_;
    hmac_sha1_sum(pack('h*',$seed), $app->secrets->[0]);
}

sub _reset_timer {
    state $hb_wait;
    state $timer;
    $hb_wait = shift if @_;
    Mojo::IOLoop->remove($timer)
	if defined $timer;
    $timer = Mojo::IOLoop->timer($hb_wait, sub { 
				     print STDERR "stopping...";
				     shift->stop;
				 });
}


1;

=encoding utf8

=head1 SYNOPSIS

  # Mojolicious::Lite

  plugin 'LocalUI';

  get '/' => "index";
  #.. and whatever else

  app->start;
  __DATA__

  @@ index.html.ep
  % layout 'default';
  ...

  @@ layouts/default.html.ep
  <!DOCTYPE html>
  <html>
    <head> ...
  %= $c->localui->jsload;
    </head>
    <body><%= content %></body>
  </html>

=head1 DESCRIPTION

On server start, L<Mojolicious::Plugin::LocalUI> opens a dedicated window in your default internet browser.  This assumes an available desktop that L<Browser::Open> knows how to deal with.  The application must be listening on a loopback/localhost port (dies otherwise), and the server shuts down when the browser window (and descendants thereof) is subsequently closed.

This is a way to create low-effort desktop applications using L<Mojolicious> (cross-platform if your code is sufficiently portable).

=head1 OPTIONS

=head2 initial_wait

How many seconds to wait on server start for browser window to finish loading.  Defaults to 10.

=head2 final_wait

How many seconds to wait after browser window ceases communicating before terminating (we do not rely on C<window.unload>).  Defaults to 3.

Since javascript timer events from backgrounded/hidden tabs/windows are typically throttled, reducing this below 2 will most likely make the application terminate whenever the window is hidden or minimized.

=head2 entry_path

URI path for the entry point of your application (i.e., what to display when the brower window opens).  Defaults to C</>.

=head2 api_prefix

Path prefix for URIs used by this plugin.  This is where the various endpoints needed by this module live (and is also where the required javascript file(s) are served from).  Defaults to C</hb/>; it can be pretty much anything as long as it's distinct from what the rest of your application uses.  But it would be best to keep it short.

=head1 HELPERS

=head2 localui.jsload

Loads whatever javascript needs to be in the <head> section of every page to be displayed in the browser window.  You most likely want this in your default layout.

  %= $c->localui->jsload;

Or you can be more elaborate

  %= javascript 'https://code.jquery.com/jquery-3.3.1.min.js';
  %= $c->localui->jsload( nojquery => 1, begin
        .on_hb(function(h) {
          // do something on every heartbeat
	  $('#heartbeat').html(h);
        })
  % end );

Options include

=head3 nojquery

Suppress loading of jquery from the Mojolicious distribution, meaning you already loaded it from elsewhere (needs to be at least version 3.0).

=head3 nofinish

Suppress the default C<on_finish> handler.

=head3 begin

Final C<begin> block, if provided, will be assumed to be javascript code to further configure the heartbeat object (code is preceded by C<$().heartbeat()>), typically to add C<on_hb> or C<on_finish> handlers

=head1 METHODS

L<Mojolicious::Plugin::LocalUI> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Mojolicious->new, entry_path => '/');

Register plugin in L<Mojolicious> application.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<https://mojolicious.org>.

=cut
__DATA__

@@ layouts/done.html.ep
<!DOCTYPE html>
<html>
  <head><title><%= title %></title>
%= stylesheet begin
body { background-color: #ddd; font-family: helvetica; }
h1 { font-size: 40px; color: white }
% end
  </head>
  <body><%= content %></body>
</html>

@@ done.html.ep
% layout 'done';
% title $title;
<h1 id="header"><%= $header %></h1>
<h2><span id="status"></span> <span id="message"></span></h2>
<p id="info"></p>

@@ ready.js.ep
$.ready.then(function() {
    $().heartbeat()
%== $_cb ? $_cb->() : ''
% unless ($nofinish) {
    .on_finish(function(unexpected,o) {
% my ($hd,$bdy) = do {
%   my $d = Mojo::DOM->new($c->render_to_string(template => "done", format => "html", title => "Finished", header => "Close this window"));
%   map {"'" . (Mojo::Util::trim($d->at($_)->content)
%                =~ s/'/\\047/gr =~ s/\n/'\n    +'\\n/gr) . "'"} qw(head body)
% };
      $('head').html(<%== $hd %>);
      $('body').html(<%== $bdy %>);
      if (unexpected) {
        $('#header').html('Error');
        $('#status').html(o.code);
        $('#message').html(o.msg);
        $('#info').html(o.status == 'error' ? '' : "("+o.msg+")");
      }
    })
% }
    .start();
});
